### PR TITLE
users can disable the gradient computation [new]

### PR DESCRIPTION
**Pull request by yorkerlin**
@karlnapf 
take a look at this

### CODE REVIEW

File: src/shogun/base/SGObject.cpp [hunk 1]
```diff
@@ -254,6 +254,7 @@ void CSGObject::update_parameter_hash()
 	uint32_t carry=0;
 	uint32_t length=0;
 
+	m_hash=0;
```
Comment (yorkerlin) [inline src/shogun/base/SGObject.cpp]: @karlnapf 
the bug is mentioned at
https://github.com/shogun-toolbox/shogun/issues/2838

Comment (karlnapf) [inline src/shogun/base/SGObject.cpp]: please send a seperate PR for this fix, and add unit tests that make sure the rationale is explicit

**Comment by karlnapf** [comment]
How does this work from the user perspective now?

**Comment by yorkerlin** [comment]
ref
https://github.com/shogun-toolbox/shogun/issues/2817

**Comment by yorkerlin** [comment]
@karlnapf 
the gradient computation will only be activated when any of the following things:
* gradients 
* posterior mean/cov  
are required

**Comment by karlnapf** [comment]
Cool thats perfpect.

As said, could you send a seperate PR with the bugfix and a unit test for the hash behavious? thanks!

**Comment by yorkerlin** [comment]
@karlnapf 
the unit tests are added.

**Comment by karlnapf** [comment]
Great thanks!

**Pull request merged**